src/GLua/AG/PrettyPrint.ag

```diff
@@ -483,7 +497,12 @@ sem FuncName
 
 sem PrefixExp
     | PFVar      lhs.pretty = tok @name <> @suffixes.pretty
-    | ExprVar    lhs.pretty = parens @lhs.ppconf NonEmpty @expr.pretty <> @suffixes.pretty
+    | ExprVar    lhs.pretty = (if @loc.noparens then @expr.pretty else parens @lhs.ppconf NonEmpty @expr.pretty) <> @suffixes.pretty
+                 lhs.precedence = if @loc.noparens then @expr.precedence else 10
+                 loc.wraps = case @expr.copy of
+                                MExpr _ (APrefixExpr _) -> True
+                                _ -> False
+                 loc.noparens = removeExtraParens @lhs.ppconf && (@loc.wraps || (length @suffixes.copy == 0 && (@expr.precedence == 10 || @lhs.neighborPrecedence == 0 || removeAllExtraParens @lhs.ppconf && @lhs.neighborPrecedence < @expr.precedence)))
```

Comment:
The parser knows 8 levels of precedence (specifically, levels 1 to 8), though here I see both 0 and 10 present. The seem to be some sorts of bottom and top values, but I'm not sure why there's a need for those otherwise out-of-range values.

It's probably better to have a data type representing levels 1 to 8, with an `Ord` instance to facilitate comparison.

Reply:
I think we need top and bottom values. The bottom value 0 is useful as a setting for the neighbor precedence of the base level of an expression where nothing can possibly "pull apart" a binary expression. The top value 10 (should be 9 I guess) is useful for things like variables which also can't ever be pulled apart, so they never need parenthesis.